pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,292 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"reflect"
+	"sync/atomic"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/util/proxy"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// informerDefaultResync is the default resync period if a handler doesn't specify one.
+	// Use the same default value as kube-controller-manager:
+	// https://github.com/kubernetes/kubernetes/blob/release-1.17/pkg/controller/apis/config/v1alpha1/defaults.go#L120
+	informerDefaultResync = 12 * time.Hour
+
+	minRetryDelay = 100 * time.Millisecond
+	maxRetryDelay = 30 * time.Second
+)
+
+// Listener defines the interface which needs to be implemented by clients which want to subscribe
+// to Endpoint updates.
+type Listener interface {
+	Enqueue()
+}
+
+// EndpointResolver is in charge of resolving a specific Service Endpoint, which can then be
+// accessed directly instead of depending on the ClusterIP functionality provided by K8s proxies
+// (whether it's kube-proxy or AntreaProxy). A new Endpoint is resolved every time the Service's
+// Spec or the Ednpoints' Subsets are updated, and registered listeners are notified. While this
+// EndpointResolver is somewhat generic, at the moment it is only meant to be used for the Antrea
+// Service.
+type EndpointResolver struct {
+	// name is the name of the controller in charge of Endpoint resolution.
+	name        string
+	namespace   string
+	serviceName string
+	servicePort int32
+	// informerFactory is stored here so it can be started in the Run() method.
+	informerFactory informers.SharedInformerFactory
+	// serviceLister is used to retrieve the Service when selecting an Endpoint.
+	serviceLister       corev1listers.ServiceLister
+	serviceListerSynced cache.InformerSynced
+	// endpointLister is used to retrieve the Endpoints for the Service during Endpoint selection.
+	endpointsLister       corev1listers.EndpointsLister
+	endpointsListerSynced cache.InformerSynced
+	queue                 workqueue.RateLimitingInterface
+	// listeners need to implement the Listerner interface and will get notified when the
+	// current Endpoint URL changes.
+	listeners   []Listener
+	endpointURL atomic.Pointer[url.URL]
+}
+
+func NewEndpointResolver(kubeClient kubernetes.Interface, namespace, serviceName string, servicePort int32) *EndpointResolver {
+	key := namespace + "/" + serviceName
+	controllerName := fmt.Sprintf("ServiceEndpointResolver-%s", key)
```

Comment:
Is `ServiceEndpointResolver:kube-system/antrea` more readable than `ServiceEndpointResolver-kube-system/antrea`?